### ISSUE TITLE
[React] Needs Clarity - What is Needed for Exercise is not Clear in the Instructions

### ISSUE THREAD

**Issue by mohmad-null** [exercism/python#2824]
There's a 15-day old Mentoring request for the React task that basically just says: "I don't understand the task, please explain". Reading the instructions for the task I have to agree; it's not at all clear what it's asking for.

I could probably decipher it from the tests, but that seems like a lot of work and evidently it's not something this user is able/willing to do either.

I'd suggest some combination of:
* Improve the instructions
* Link to a web-page that clearly explains what a "reactive system" is.

(There are a number of tasks which seem to require domain knowledge/expertise to be able to solve. I'm not sure this is really something that's desirable on a site designed to help folks learn a language)

**Comment by BethanyG** [comment]
Hi @mohmad-null! 

First, _thank you_ 💜   for investigating why a student is stuck, and thinking through what might help them be less stuck!

Problem descriptions for practice exercises are drawn from a central repository called [problem specifications](https://github.com/exercism/problem-specifications).  Often, tests for practice exercises are pulled from this repository as well - although in this case, our `React` test file is not auto-generated.

While we could add a Python-specific addendum, I think it might be best to bring up your concerns in problem-specifications first, to see if a proposed change in instructions could help for more than just the Python track.

As for the student "not wanting" or being "unable" to figure out requirements from tests -- it can be difficult and/or frustrating at times, and it is why we are considering having a queue for students who are really stuck (_they can't pass all the tests on their own_) so they can get targeted help from a mentor.

However, the overriding principal of practice exercises is less _teaching_ and more _practice_.   They are `TDD` -- the students are _expected_ to have all tests fail at first, then proceed to write code until the tests pass.  They're expected to look at the test files, should they need additional "specs" over the exercise instructions.  Instructions are kept somewhat general, to encourage as much freedom as possible in implementations -- not every student is going to want to practice "reactive" programming the same way -- Some may want to use RxPy, some may want to use pipelining via generators or do something from scratch.  Some approaches might be fully functional/declarative, some more OOP.  It's up to the student at this level to decide on an approach, and then refine that approach via mentor conversations. 

The one caveat I would put here is that if a student were to use `RxPy`, we would need to account for that in the website docker containers , since third-party libraries need to be pre-loaded -- but that's a different conversation for a different time.  😄

**Comment by mohmad-null** [comment]
@BethanyG - Thanks for your feedback. Raised here: https://github.com/exercism/problem-specifications/issues/1902

> They are TDD -- the students are expected to have all tests fail at first, then proceed to write code until the tests pass. They're expected to look at the test files, should they need additional "specs" over the exercise instructions.

I get what you're saying but fervently believe that you can't actually solve many problems without understanding the problem in the first place. Devs who only mindlessly code-to-spec aren't really something you want.
Lets say there was a exercise for doing Fibonacci sequence's. "Find the nth character" and there are a few tests up to `n=20`. If the Fibonacci isn't explained in the blurb, the simple solution to pass the tests is to just hard-code a list with the first 20 values in the Fibonacci sequence and return the index. This isn't "wrong" - the Fibonacci doesn't change after all - but for our purposes (learning) it's sub-optimal compared to using the sequence.

>  It's up to the student at this level to decide on an approach
This I completely agree with. (Well, I mean I would if I had any clue what a "react system" was and therefore what level of complexity a result would entail :-) ). But as with my Fibonacci example above, a user will have no clue what the possibilities are if they don't know about how the Fibonacci sequence is calculated.

TL;DR: It's impossible to evaluate solutions until you understand the problem-space.

**Comment by BethanyG** [comment]
> I get what you're saying but fervently believe that you can't actually solve many problems without understanding the problem in the first place. Devs who only mindlessly code-to-spec aren't really something you want.

_But isn't over-specifying implementation expecting exactly that -- that someone blindly follow a spec instead of working through what they are given and researching what they need to do?_  It's a hard balance between describing/giving detail and implying a strategy via step-by-step direction.

Take your Fibonacci example -- I can link to good definitions.  I can draw a diagram, or show examples.  **_However_** -- none of those things are going to help you with what is really going on with most Fibonacci-related problems in Python:  calculating the sequence from the beginning to any significant length will run you out of memory, so a programmer has to figure out how they will manage that scenario.  And it is _frustrating_.  It's a similar scenario to that massive test case for `Pythagorean Triplet`.  

But if I give you links to memoization examples and other strategies and answers to that -- I am both denying you the experience of learning and the pride that comes when you work (with or without mentor help) out a solution.

As for the developer that hard-codes the sequence to pass the tests?  Yes -- that's an answer.  I've done that with some problems. We're not here to dictate how someone solves a problem or how they go about learning.  It's on them.  If I were that person's mentor, I would encourage them to stretch themselves, and challenge them to find a more general or programmatic solution -- but I'm not here to give grades, certifications or lectures.  I'm here to provide programming problems with well-vetted test suites and mentorship/code feedback.  If what they choose to do is simply hard-code answers ... they're allowed to do that, and they are allowed to skip or give up what they otherwise might learn.

I think finding this sort of balance between giving a student enough information that they feel empowered to take on a programming problem _without_ crushing them under requirements and specifications is quite challenging.  Especially when you are trying to challenge programmers at many different levels coming from many different perspectives.  What is under-specification to an "intermediate" programmer can be massive over-specification to an "advanced" one - and total gibberish to a "beginner".  And that's _without_ the cultural, linguistic, and economic complications inherent in the problem presentations.

**Comment by BethanyG** [comment]
Closing this in favor of [exercism/problem-specifications#1902](https://github.com/exercism/problem-specifications/issues/1902)
